From 824331eefa9f0b2b8485cc978be945b43426388e Mon Sep 17 00:00:00 2001
From: Jouni Malinen <j@w1.fi>
Date: Sun, 25 Oct 2015 15:45:50 +0200
Subject: [PATCH 08/12] WNM: Ignore Key Data in WNM Sleep Mode Response frame
 if no PMF in use

WNM Sleep Mode Response frame is used to update GTK/IGTK only if PMF is
enabled. Verify that PMF is in use before using this field on station
side to avoid accepting unauthenticated key updates. (CVE-2015-5310)

Signed-off-by: Jouni Malinen <j@w1.fi>
---
 wpa_supplicant/wnm_sta.c | 6 ++++++
 1 file changed, 6 insertions(+)

Index: wpa-2.4/wpa_supplicant/wnm_sta.c
===================================================================
--- wpa-2.4.orig/wpa_supplicant/wnm_sta.c	2017-10-16 07:54:58.952405469 -0400
+++ wpa-2.4/wpa_supplicant/wnm_sta.c	2017-10-16 07:54:58.944405389 -0400
@@ -193,6 +193,12 @@ static void wnm_sleep_mode_exit_success(
 		return;
 	}
 
+	if (key_len_total && !wpa_sm_pmf_enabled(wpa_s->wpa)) {
+		wpa_msg(wpa_s, MSG_INFO,
+			"WNM: Ignore Key Data in WNM-Sleep Mode Response - PMF not enabled");
+		return;
+	}
+
 	while (ptr + 1 < end) {
 		if (ptr + 2 + ptr[1] > end) {
 			wpa_printf(MSG_DEBUG, "WNM: Invalid Key Data element "
